File: Lab07/log_probability.pl
#!/usr/bin/perl -w

if(@ARGV != 1) {           #error checking
   die "Usage: $0 <word>\n";
}

$searchWord = lc $ARGV[0];

foreach $file (glob "poets/*.txt") {     #loop through all file names with .txt 
   my $poetName = $file;   
   $poetName =~ s/.*\///;     #getting poet name from file path 
   $poetName =~ s/\.txt$//;
   $poetName =~ s/_/ /g;   
   
   $wordCount{$poetName} = 0;    #initalising, used for 0 case
   
   open poetFile, "<$file" or die;
   while($line = <poetFile>) {
         foreach my $word ($line =~ /[a-z]+/ig) {  #capture all occurences, case insensitive
            $word = lc $word;
            $totalWords{$poetName}++;           #incrementing hashes
            $wordCount{$poetName}++ if $word eq $searchWord;
         }
   }   
}

foreach my $poet (sort keys %wordCount) {
   printf "log((%d+1)/%6d) = %8.4f %s\n", $wordCount{$poet}, $totalWords{$poet}, 
         log(($wordCount{$poet}+1)/$totalWords{$poet}), $poet;

}

File: Lab07/log_probability.py
#!/usr/bin/python

import sys, glob, re, math

if len(sys.argv) != 2:
   print "Usage %s <word>" %sys.argv[0]
   sys.exit(1)
   
searchWord = sys.argv[1].lower()

wordCount = {}
totalWords = {}

for poetFile in glob.glob("poets/*.txt"):
   poetName = re.sub(r'.*/','',poetFile)
   poetName = re.sub(r'.txt','',poetName)
   poetName = re.sub(r'_',' ',poetName)
   
   wordCount[poetName] = 0
   totalWords[poetName] = 0
   
   for line in open(poetFile): 
      
      for word in re.findall(r'[a-zA-Z]+',line):   
         word = word.lower()
         totalWords[poetName] += 1
      
         if word == searchWord:
            wordCount[poetName] += 1

for poet in sorted(wordCount):
   print "log((%d+1)/%6d) = %8.4f %s" % (wordCount[poet], totalWords[poet], 
                                          math.log((wordCount[poet]+1)/float(totalWords[poet])), poet)
